### PR TITLE
syntax and basic language description

### PR DESCRIPTION
**Pull request by franziskuskiefer**
Let's get a concise description of the language into the book to replace the `Language.md` from the main repo, have a basis for discussions, and a starting point for new backends.
This is just taken from the paper, which is a little more up to date than the `Language.md` but still not correct.

@denismerigoux @cmester0 Could you make this correct and add a little prose?

### CODE REVIEW

**Comment by spitters** [comment]
Perhaps we may want to specify hacspec in ott at some point. This would
give us a definition of the language in various backends.
This would be a step in proving soundness of the operational semantics wrt
the denotational semantics.
I'm not sure how much time that would take.
https://github.com/ott-lang/ott

File: src/language/syntax.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    | e                         Return expression
+    | b                         Statement block
+e ::= ()|true|false             Unit and boolean literals
+    | n                         Integer literal
+    | x                         Variable
+    | f([a]+)                   Function call
+    | e ⊙ e                     Binary operations
+    | ⊘ e                       Unary operations
+    | ([e]+)                    Tuple constructor
+    | e.(n)                     Tuple field access where n is a natural number
```
Comment (franziskuskiefer) [inline src/language/syntax.md:27]: I don't think this is actually implemented @denismerigoux ?

File: src/language/syntax.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+```
+P ::= [i]\*                     Program items
+i ::= array!(t, μ, n)           Array type declaration where n is a natural number
+    | nat_mod!(t, n)            Modular integer
+    | fn f([d]+) -> μ b         Function declaration
+    | type t = enum { [c(μ)]+ } Enum declaration (with constructors)
+    | type t = struct([μ]+)     Struct declaration (without named fields)
+    | const x: μ = e            Constant declaration
+    | use crate::*              External crate import
```
Comment (franziskuskiefer) [inline src/language/syntax.md:11]: We also have `use mod::*` now.

File: src/language/syntax.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+d ::= x: τ                      Function argument
+μ ::= unit|bool|u8|U8|i8|I8...  Base types
```
Comment (franziskuskiefer) [inline src/language/syntax.md:13]: We have to define the `U8...` (secret integers) somewhere.

File: src/language/syntax.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+b ::= {[s;]+}                   Block
+p ::= x                         Variable pattern
+    | ([p]+)                    Tuple pattern
+    | t (p)                     Struct constructor pattern
+    | _                         Wildcard pattern
+s ::= let p: τ = e              Let binding
+    | let mut p: τ = e          Mutable let binding
+    | x = e                     Variable reassignment
+    | x = e?                    Variable reassignment with error catching
+    | let p: τ = e?             Let binding with error catching
```
Comment (franziskuskiefer) [inline src/language/syntax.md:30]: `?` catches options as well.

File: src/language/syntax.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    | x                         Variable
+    | ()                        Unit
+    | f([a]+)                   Function call
+    | if e then e else e        Conditional expression
+    | e ⊙ e                     Binary operations
+    | ⊘ e                       Unary operations
+    | ([e]+)                    Tuple constructor
+    | x[e]                      Array or sequence index
+a ::= e                         Linear argument
+    | &e                        Call-site borrowing
```
Comment (franziskuskiefer) [inline src/language/syntax.md:51]: We also allow `mut &e` in function arguments.

File: src/language/syntax.md (hunk 3 above)
Comment (denismerigoux) [reply src/language/syntax.md:13]: They are handled internally in the language as regular functions, but I can add them to the syntax.

**Comment by franziskuskiefer** [comment]
If there are no comments @cmester0, I'll go ahead and merge this.

**Comment by cmester0** [comment]
Seems good to me.

**Pull request merged**